dali/kernels/erase/erase_gpu.h

```diff
@@ -0,0 +1,526 @@
+    erase_sample_desc<T, ndim> sample, ibox<ndim> region, span<T> fill_values = {},
+    span<ibox<ndim>> erase_regions = {}, ivec<ndim> coordinate = {}, int64_t offset_base = 0) {
+  constexpr int d = current_dim;
+  constexpr int dC = d + 1;
+  int boundary = ::min(region.hi[d], sample.sample_shape[d]);
+  int boundary_c = ::min(region.hi[dC], sample.sample_shape[dC]);
+
+  for (int idx = region.lo[d] + threadIdx.y * blockDim.x + threadIdx.x; idx < boundary;
+       idx += blockDim.y * blockDim.x) {
+    coordinate[d] = idx;
+    int64_t offset = offset_base + idx * sample.sample_stride[d];
+    for (int c = region.lo[dC]; c < boundary_c; c++) {
+      coordinate[dC] = c;
+      Worker::template copy_or_erase<channel_dim>(sample, erase_regions, coordinate, offset + c,
+                                                  fill_values);
+    }
+  }
+}
+
+
+struct do_copy {
+  template <int channel_dim, typename T, int ndim>
+  __device__ static void copy_or_erase(erase_sample_desc<T, ndim> sample,
+                                       span<ibox<ndim>> erase_regions, ivec<ndim> coordinate,
+                                       int64_t offset, span<T> fill_values) {
+    (void)erase_regions;
+    (void)coordinate;
+    (void)fill_values;
+    sample.out[offset] = sample.in[offset];
+  }
+};
+
+struct do_erase {
+  template <int channel_dim, typename T, int ndim>
+  __device__ static void copy_or_erase(erase_sample_desc<T, ndim> sample,
+                                       span<ibox<ndim>> erase_regions, ivec<ndim> coordinate,
+                                       int64_t offset, span<T> fill_values) {
+    (void)erase_regions;
+    if (channel_dim == -1) {
+      sample.out[offset] = fill_values[0];
+    } else {
+      sample.out[offset] = fill_values[coordinate[channel_dim]];
+    }
+  }
+};
+
+struct do_copy_or_erase {
+  template <int channel_dim, typename T, int ndim>
+  __device__ static void copy_or_erase(erase_sample_desc<T, ndim> sample,
+                                       span<ibox<ndim>> erase_regions, ivec<ndim> coordinate,
+                                       int64_t offset, span<T> fill_values) {
+    auto fill_value = channel_dim == -1 ? fill_values[0] : fill_values[coordinate[channel_dim]];
+    copy_or_erase<channel_dim>(sample, erase_regions, coordinate, offset, fill_value);
+  }
+
+  template <int channel_dim, typename T, int ndim>
+  __device__ static void copy_or_erase(erase_sample_desc<T, ndim> sample,
+                                       span<ibox<ndim>> erase_regions, ivec<ndim> coordinate,
+                                       int64_t offset, T fill_value) {
+    bool copy = true;
+    for (auto &region : erase_regions) {
+      if (region.contains(coordinate)) {
+        copy = false;
+      }
+    }
+    if (copy) {
+      sample.out[offset] = sample.in[offset];
+    } else {
+      sample.out[offset] = fill_value;
```

Comment:
I don't know how smart NVCC is, but it'd say that this:
```c++
sample.out[offset] = copy ? sample.in[offset] : fill_value;
```
has fewer opportunities for the compiler to screw up optimization and write coalescing.

Reply:
Looking how well it handles vec & Box, I would say it's quite good.

Reply:
Done regardless